// @aws-sdk/client-ses/dist-es/commands/CreateConfigurationSetEventDestinationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateConfigurationSetEventDestinationRequest, CreateConfigurationSetEventDestinationResponse, } from "../models/models_0";
import { deserializeAws_queryCreateConfigurationSetEventDestinationCommand, serializeAws_queryCreateConfigurationSetEventDestinationCommand, } from "../protocols/Aws_query";
var CreateConfigurationSetEventDestinationCommand = (function (_super) {
    __extends(CreateConfigurationSetEventDestinationCommand, _super);
    function CreateConfigurationSetEventDestinationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateConfigurationSetEventDestinationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "CreateConfigurationSetEventDestinationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateConfigurationSetEventDestinationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateConfigurationSetEventDestinationResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateConfigurationSetEventDestinationCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateConfigurationSetEventDestinationCommand(input, context);
    };
    CreateConfigurationSetEventDestinationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateConfigurationSetEventDestinationCommand(output, context);
    };
    return CreateConfigurationSetEventDestinationCommand;
}($Command));
export { CreateConfigurationSetEventDestinationCommand };
