// @aws-sdk/client-ses/dist-es/commands/CreateTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateTemplateRequest, CreateTemplateResponse } from "../models/models_0";
import { deserializeAws_queryCreateTemplateCommand, serializeAws_queryCreateTemplateCommand, } from "../protocols/Aws_query";
var CreateTemplateCommand = (function (_super) {
    __extends(CreateTemplateCommand, _super);
    function CreateTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "CreateTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateTemplateCommand(input, context);
    };
    CreateTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateTemplateCommand(output, context);
    };
    return CreateTemplateCommand;
}($Command));
export { CreateTemplateCommand };
