// @aws-sdk/client-ses/dist-es/commands/CreateReceiptRuleSetCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateReceiptRuleSetRequest, CreateReceiptRuleSetResponse } from "../models/models_0";
import { deserializeAws_queryCreateReceiptRuleSetCommand, serializeAws_queryCreateReceiptRuleSetCommand, } from "../protocols/Aws_query";
var CreateReceiptRuleSetCommand = (function (_super) {
    __extends(CreateReceiptRuleSetCommand, _super);
    function CreateReceiptRuleSetCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateReceiptRuleSetCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "CreateReceiptRuleSetCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateReceiptRuleSetRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateReceiptRuleSetResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateReceiptRuleSetCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateReceiptRuleSetCommand(input, context);
    };
    CreateReceiptRuleSetCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateReceiptRuleSetCommand(output, context);
    };
    return CreateReceiptRuleSetCommand;
}($Command));
export { CreateReceiptRuleSetCommand };
