// @aws-sdk/client-ses/dist-es/commands/CreateCustomVerificationEmailTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateCustomVerificationEmailTemplateRequest } from "../models/models_0";
import { deserializeAws_queryCreateCustomVerificationEmailTemplateCommand, serializeAws_queryCreateCustomVerificationEmailTemplateCommand, } from "../protocols/Aws_query";
var CreateCustomVerificationEmailTemplateCommand = (function (_super) {
    __extends(CreateCustomVerificationEmailTemplateCommand, _super);
    function CreateCustomVerificationEmailTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateCustomVerificationEmailTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "CreateCustomVerificationEmailTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateCustomVerificationEmailTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateCustomVerificationEmailTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateCustomVerificationEmailTemplateCommand(input, context);
    };
    CreateCustomVerificationEmailTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateCustomVerificationEmailTemplateCommand(output, context);
    };
    return CreateCustomVerificationEmailTemplateCommand;
}($Command));
export { CreateCustomVerificationEmailTemplateCommand };
