// @aws-sdk/client-ses/dist-es/commands/CreateReceiptFilterCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateReceiptFilterRequest, CreateReceiptFilterResponse } from "../models/models_0";
import { deserializeAws_queryCreateReceiptFilterCommand, serializeAws_queryCreateReceiptFilterCommand, } from "../protocols/Aws_query";
var CreateReceiptFilterCommand = (function (_super) {
    __extends(CreateReceiptFilterCommand, _super);
    function CreateReceiptFilterCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateReceiptFilterCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "CreateReceiptFilterCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateReceiptFilterRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateReceiptFilterResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateReceiptFilterCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateReceiptFilterCommand(input, context);
    };
    CreateReceiptFilterCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateReceiptFilterCommand(output, context);
    };
    return CreateReceiptFilterCommand;
}($Command));
export { CreateReceiptFilterCommand };
