// @aws-sdk/client-ses/dist-es/commands/DeleteIdentityCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteIdentityRequest, DeleteIdentityResponse } from "../models/models_0";
import { deserializeAws_queryDeleteIdentityCommand, serializeAws_queryDeleteIdentityCommand, } from "../protocols/Aws_query";
var DeleteIdentityCommand = (function (_super) {
    __extends(DeleteIdentityCommand, _super);
    function DeleteIdentityCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteIdentityCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DeleteIdentityCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteIdentityRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteIdentityResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteIdentityCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteIdentityCommand(input, context);
    };
    DeleteIdentityCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteIdentityCommand(output, context);
    };
    return DeleteIdentityCommand;
}($Command));
export { DeleteIdentityCommand };
