// @aws-sdk/client-ses/dist-es/commands/DeleteReceiptRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteReceiptRuleRequest, DeleteReceiptRuleResponse } from "../models/models_0";
import { deserializeAws_queryDeleteReceiptRuleCommand, serializeAws_queryDeleteReceiptRuleCommand, } from "../protocols/Aws_query";
var DeleteReceiptRuleCommand = (function (_super) {
    __extends(DeleteReceiptRuleCommand, _super);
    function DeleteReceiptRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteReceiptRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DeleteReceiptRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteReceiptRuleRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteReceiptRuleResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteReceiptRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteReceiptRuleCommand(input, context);
    };
    DeleteReceiptRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteReceiptRuleCommand(output, context);
    };
    return DeleteReceiptRuleCommand;
}($Command));
export { DeleteReceiptRuleCommand };
