// @aws-sdk/client-ses/dist-es/commands/DeleteConfigurationSetEventDestinationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteConfigurationSetEventDestinationRequest, DeleteConfigurationSetEventDestinationResponse, } from "../models/models_0";
import { deserializeAws_queryDeleteConfigurationSetEventDestinationCommand, serializeAws_queryDeleteConfigurationSetEventDestinationCommand, } from "../protocols/Aws_query";
var DeleteConfigurationSetEventDestinationCommand = (function (_super) {
    __extends(DeleteConfigurationSetEventDestinationCommand, _super);
    function DeleteConfigurationSetEventDestinationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteConfigurationSetEventDestinationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DeleteConfigurationSetEventDestinationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteConfigurationSetEventDestinationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteConfigurationSetEventDestinationResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteConfigurationSetEventDestinationCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteConfigurationSetEventDestinationCommand(input, context);
    };
    DeleteConfigurationSetEventDestinationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteConfigurationSetEventDestinationCommand(output, context);
    };
    return DeleteConfigurationSetEventDestinationCommand;
}($Command));
export { DeleteConfigurationSetEventDestinationCommand };
