// @aws-sdk/client-ses/dist-es/commands/DeleteReceiptFilterCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteReceiptFilterRequest, DeleteReceiptFilterResponse } from "../models/models_0";
import { deserializeAws_queryDeleteReceiptFilterCommand, serializeAws_queryDeleteReceiptFilterCommand, } from "../protocols/Aws_query";
var DeleteReceiptFilterCommand = (function (_super) {
    __extends(DeleteReceiptFilterCommand, _super);
    function DeleteReceiptFilterCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteReceiptFilterCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DeleteReceiptFilterCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteReceiptFilterRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteReceiptFilterResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteReceiptFilterCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteReceiptFilterCommand(input, context);
    };
    DeleteReceiptFilterCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteReceiptFilterCommand(output, context);
    };
    return DeleteReceiptFilterCommand;
}($Command));
export { DeleteReceiptFilterCommand };
