// @aws-sdk/client-ses/dist-es/commands/DeleteConfigurationSetTrackingOptionsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteConfigurationSetTrackingOptionsRequest, DeleteConfigurationSetTrackingOptionsResponse, } from "../models/models_0";
import { deserializeAws_queryDeleteConfigurationSetTrackingOptionsCommand, serializeAws_queryDeleteConfigurationSetTrackingOptionsCommand, } from "../protocols/Aws_query";
var DeleteConfigurationSetTrackingOptionsCommand = (function (_super) {
    __extends(DeleteConfigurationSetTrackingOptionsCommand, _super);
    function DeleteConfigurationSetTrackingOptionsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteConfigurationSetTrackingOptionsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DeleteConfigurationSetTrackingOptionsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteConfigurationSetTrackingOptionsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteConfigurationSetTrackingOptionsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteConfigurationSetTrackingOptionsCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteConfigurationSetTrackingOptionsCommand(input, context);
    };
    DeleteConfigurationSetTrackingOptionsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteConfigurationSetTrackingOptionsCommand(output, context);
    };
    return DeleteConfigurationSetTrackingOptionsCommand;
}($Command));
export { DeleteConfigurationSetTrackingOptionsCommand };
