// @aws-sdk/client-ses/dist-es/commands/DescribeActiveReceiptRuleSetCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeActiveReceiptRuleSetRequest, DescribeActiveReceiptRuleSetResponse } from "../models/models_0";
import { deserializeAws_queryDescribeActiveReceiptRuleSetCommand, serializeAws_queryDescribeActiveReceiptRuleSetCommand, } from "../protocols/Aws_query";
var DescribeActiveReceiptRuleSetCommand = (function (_super) {
    __extends(DescribeActiveReceiptRuleSetCommand, _super);
    function DescribeActiveReceiptRuleSetCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeActiveReceiptRuleSetCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "DescribeActiveReceiptRuleSetCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeActiveReceiptRuleSetRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeActiveReceiptRuleSetResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeActiveReceiptRuleSetCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDescribeActiveReceiptRuleSetCommand(input, context);
    };
    DescribeActiveReceiptRuleSetCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDescribeActiveReceiptRuleSetCommand(output, context);
    };
    return DescribeActiveReceiptRuleSetCommand;
}($Command));
export { DescribeActiveReceiptRuleSetCommand };
