// @aws-sdk/client-ses/dist-es/commands/GetAccountSendingEnabledCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetAccountSendingEnabledResponse } from "../models/models_0";
import { deserializeAws_queryGetAccountSendingEnabledCommand, serializeAws_queryGetAccountSendingEnabledCommand, } from "../protocols/Aws_query";
var GetAccountSendingEnabledCommand = (function (_super) {
    __extends(GetAccountSendingEnabledCommand, _super);
    function GetAccountSendingEnabledCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetAccountSendingEnabledCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "GetAccountSendingEnabledCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: function (input) { return input; },
            outputFilterSensitiveLog: GetAccountSendingEnabledResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetAccountSendingEnabledCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryGetAccountSendingEnabledCommand(input, context);
    };
    GetAccountSendingEnabledCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryGetAccountSendingEnabledCommand(output, context);
    };
    return GetAccountSendingEnabledCommand;
}($Command));
export { GetAccountSendingEnabledCommand };
