// @aws-sdk/client-ses/dist-es/commands/GetIdentityDkimAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetIdentityDkimAttributesRequest, GetIdentityDkimAttributesResponse } from "../models/models_0";
import { deserializeAws_queryGetIdentityDkimAttributesCommand, serializeAws_queryGetIdentityDkimAttributesCommand, } from "../protocols/Aws_query";
var GetIdentityDkimAttributesCommand = (function (_super) {
    __extends(GetIdentityDkimAttributesCommand, _super);
    function GetIdentityDkimAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetIdentityDkimAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "GetIdentityDkimAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetIdentityDkimAttributesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetIdentityDkimAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetIdentityDkimAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryGetIdentityDkimAttributesCommand(input, context);
    };
    GetIdentityDkimAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryGetIdentityDkimAttributesCommand(output, context);
    };
    return GetIdentityDkimAttributesCommand;
}($Command));
export { GetIdentityDkimAttributesCommand };
