// @aws-sdk/client-ses/dist-es/commands/GetTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetTemplateRequest, GetTemplateResponse } from "../models/models_0";
import { deserializeAws_queryGetTemplateCommand, serializeAws_queryGetTemplateCommand } from "../protocols/Aws_query";
var GetTemplateCommand = (function (_super) {
    __extends(GetTemplateCommand, _super);
    function GetTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "GetTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryGetTemplateCommand(input, context);
    };
    GetTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryGetTemplateCommand(output, context);
    };
    return GetTemplateCommand;
}($Command));
export { GetTemplateCommand };
