// @aws-sdk/client-ses/dist-es/commands/ListVerifiedEmailAddressesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListVerifiedEmailAddressesResponse } from "../models/models_0";
import { deserializeAws_queryListVerifiedEmailAddressesCommand, serializeAws_queryListVerifiedEmailAddressesCommand, } from "../protocols/Aws_query";
var ListVerifiedEmailAddressesCommand = (function (_super) {
    __extends(ListVerifiedEmailAddressesCommand, _super);
    function ListVerifiedEmailAddressesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListVerifiedEmailAddressesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "ListVerifiedEmailAddressesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: function (input) { return input; },
            outputFilterSensitiveLog: ListVerifiedEmailAddressesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListVerifiedEmailAddressesCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryListVerifiedEmailAddressesCommand(input, context);
    };
    ListVerifiedEmailAddressesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryListVerifiedEmailAddressesCommand(output, context);
    };
    return ListVerifiedEmailAddressesCommand;
}($Command));
export { ListVerifiedEmailAddressesCommand };
