// @aws-sdk/client-ses/dist-es/commands/SendBulkTemplatedEmailCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendBulkTemplatedEmailRequest, SendBulkTemplatedEmailResponse } from "../models/models_0";
import { deserializeAws_querySendBulkTemplatedEmailCommand, serializeAws_querySendBulkTemplatedEmailCommand, } from "../protocols/Aws_query";
var SendBulkTemplatedEmailCommand = (function (_super) {
    __extends(SendBulkTemplatedEmailCommand, _super);
    function SendBulkTemplatedEmailCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SendBulkTemplatedEmailCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "SendBulkTemplatedEmailCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SendBulkTemplatedEmailRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SendBulkTemplatedEmailResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SendBulkTemplatedEmailCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySendBulkTemplatedEmailCommand(input, context);
    };
    SendBulkTemplatedEmailCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySendBulkTemplatedEmailCommand(output, context);
    };
    return SendBulkTemplatedEmailCommand;
}($Command));
export { SendBulkTemplatedEmailCommand };
