// @aws-sdk/client-ses/dist-es/commands/SendRawEmailCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendRawEmailRequest, SendRawEmailResponse } from "../models/models_0";
import { deserializeAws_querySendRawEmailCommand, serializeAws_querySendRawEmailCommand } from "../protocols/Aws_query";
var SendRawEmailCommand = (function (_super) {
    __extends(SendRawEmailCommand, _super);
    function SendRawEmailCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SendRawEmailCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "SendRawEmailCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SendRawEmailRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SendRawEmailResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SendRawEmailCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySendRawEmailCommand(input, context);
    };
    SendRawEmailCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySendRawEmailCommand(output, context);
    };
    return SendRawEmailCommand;
}($Command));
export { SendRawEmailCommand };
