// @aws-sdk/client-ses/dist-es/commands/SendTemplatedEmailCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendTemplatedEmailRequest, SendTemplatedEmailResponse } from "../models/models_0";
import { deserializeAws_querySendTemplatedEmailCommand, serializeAws_querySendTemplatedEmailCommand, } from "../protocols/Aws_query";
var SendTemplatedEmailCommand = (function (_super) {
    __extends(SendTemplatedEmailCommand, _super);
    function SendTemplatedEmailCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SendTemplatedEmailCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "SendTemplatedEmailCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SendTemplatedEmailRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SendTemplatedEmailResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SendTemplatedEmailCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySendTemplatedEmailCommand(input, context);
    };
    SendTemplatedEmailCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySendTemplatedEmailCommand(output, context);
    };
    return SendTemplatedEmailCommand;
}($Command));
export { SendTemplatedEmailCommand };
