// @aws-sdk/client-ses/dist-es/commands/SetIdentityDkimEnabledCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SetIdentityDkimEnabledRequest, SetIdentityDkimEnabledResponse } from "../models/models_0";
import { deserializeAws_querySetIdentityDkimEnabledCommand, serializeAws_querySetIdentityDkimEnabledCommand, } from "../protocols/Aws_query";
var SetIdentityDkimEnabledCommand = (function (_super) {
    __extends(SetIdentityDkimEnabledCommand, _super);
    function SetIdentityDkimEnabledCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SetIdentityDkimEnabledCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "SetIdentityDkimEnabledCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SetIdentityDkimEnabledRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SetIdentityDkimEnabledResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SetIdentityDkimEnabledCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySetIdentityDkimEnabledCommand(input, context);
    };
    SetIdentityDkimEnabledCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySetIdentityDkimEnabledCommand(output, context);
    };
    return SetIdentityDkimEnabledCommand;
}($Command));
export { SetIdentityDkimEnabledCommand };
