// @aws-sdk/client-ses/dist-es/commands/UpdateReceiptRuleCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateReceiptRuleRequest, UpdateReceiptRuleResponse } from "../models/models_0";
import { deserializeAws_queryUpdateReceiptRuleCommand, serializeAws_queryUpdateReceiptRuleCommand, } from "../protocols/Aws_query";
var UpdateReceiptRuleCommand = (function (_super) {
    __extends(UpdateReceiptRuleCommand, _super);
    function UpdateReceiptRuleCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateReceiptRuleCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "UpdateReceiptRuleCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateReceiptRuleRequest.filterSensitiveLog,
            outputFilterSensitiveLog: UpdateReceiptRuleResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateReceiptRuleCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryUpdateReceiptRuleCommand(input, context);
    };
    UpdateReceiptRuleCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryUpdateReceiptRuleCommand(output, context);
    };
    return UpdateReceiptRuleCommand;
}($Command));
export { UpdateReceiptRuleCommand };
