// @aws-sdk/client-ses/dist-es/commands/UpdateAccountSendingEnabledCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateAccountSendingEnabledRequest } from "../models/models_0";
import { deserializeAws_queryUpdateAccountSendingEnabledCommand, serializeAws_queryUpdateAccountSendingEnabledCommand, } from "../protocols/Aws_query";
var UpdateAccountSendingEnabledCommand = (function (_super) {
    __extends(UpdateAccountSendingEnabledCommand, _super);
    function UpdateAccountSendingEnabledCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateAccountSendingEnabledCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "UpdateAccountSendingEnabledCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateAccountSendingEnabledRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateAccountSendingEnabledCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryUpdateAccountSendingEnabledCommand(input, context);
    };
    UpdateAccountSendingEnabledCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryUpdateAccountSendingEnabledCommand(output, context);
    };
    return UpdateAccountSendingEnabledCommand;
}($Command));
export { UpdateAccountSendingEnabledCommand };
