// @aws-sdk/client-ses/dist-es/commands/UpdateConfigurationSetReputationMetricsEnabledCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateConfigurationSetReputationMetricsEnabledRequest } from "../models/models_0";
import { deserializeAws_queryUpdateConfigurationSetReputationMetricsEnabledCommand, serializeAws_queryUpdateConfigurationSetReputationMetricsEnabledCommand, } from "../protocols/Aws_query";
var UpdateConfigurationSetReputationMetricsEnabledCommand = (function (_super) {
    __extends(UpdateConfigurationSetReputationMetricsEnabledCommand, _super);
    function UpdateConfigurationSetReputationMetricsEnabledCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateConfigurationSetReputationMetricsEnabledCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "UpdateConfigurationSetReputationMetricsEnabledCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateConfigurationSetReputationMetricsEnabledRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateConfigurationSetReputationMetricsEnabledCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryUpdateConfigurationSetReputationMetricsEnabledCommand(input, context);
    };
    UpdateConfigurationSetReputationMetricsEnabledCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryUpdateConfigurationSetReputationMetricsEnabledCommand(output, context);
    };
    return UpdateConfigurationSetReputationMetricsEnabledCommand;
}($Command));
export { UpdateConfigurationSetReputationMetricsEnabledCommand };
