// @aws-sdk/client-ses/dist-es/commands/UpdateTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateTemplateRequest, UpdateTemplateResponse } from "../models/models_0";
import { deserializeAws_queryUpdateTemplateCommand, serializeAws_queryUpdateTemplateCommand, } from "../protocols/Aws_query";
var UpdateTemplateCommand = (function (_super) {
    __extends(UpdateTemplateCommand, _super);
    function UpdateTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "UpdateTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: UpdateTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryUpdateTemplateCommand(input, context);
    };
    UpdateTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryUpdateTemplateCommand(output, context);
    };
    return UpdateTemplateCommand;
}($Command));
export { UpdateTemplateCommand };
