// @aws-sdk/client-ses/dist-es/commands/VerifyDomainDkimCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { VerifyDomainDkimRequest, VerifyDomainDkimResponse } from "../models/models_0";
import { deserializeAws_queryVerifyDomainDkimCommand, serializeAws_queryVerifyDomainDkimCommand, } from "../protocols/Aws_query";
var VerifyDomainDkimCommand = (function (_super) {
    __extends(VerifyDomainDkimCommand, _super);
    function VerifyDomainDkimCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    VerifyDomainDkimCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "VerifyDomainDkimCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: VerifyDomainDkimRequest.filterSensitiveLog,
            outputFilterSensitiveLog: VerifyDomainDkimResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    VerifyDomainDkimCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryVerifyDomainDkimCommand(input, context);
    };
    VerifyDomainDkimCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryVerifyDomainDkimCommand(output, context);
    };
    return VerifyDomainDkimCommand;
}($Command));
export { VerifyDomainDkimCommand };
